### ISSUE TITLE
revisit using cached tasks in PRs

### ISSUE THREAD

**Issue by bhearsum** [taskcluster/taskgraph#359]
Our [cached tasks implementation explicitly refuses to replace tasks with tasks from the level 1 index](https://github.com/taskcluster/taskgraph/blame/01ba699bc214ddf323db02a59cdaacaf66c41dc2/src/taskgraph/util/cached_tasks.py#L54). The commit comment as well as [original phabricator revision](https://phabricator.services.mozilla.com/D42343) talk about chain of trust not handling "tasks built from commits not at the tip of a pull requests well" as the reason for this.

Up until now this hasn't been much of an issue, as it's fairly rare to have long lived pull requests that could benefit from re-using tasks in prior pushes to the PR. In [firefox translations](https://github.com/mozilla/firefox-translations-training), this is a bit different - where some PR tasks can both be lengthy and expensive. (Arguably any project that has PR tasks that run for more than a few minutes could benefit though...)

I suggest that we find out whether or not CoT still has the aforementioned issue, and fix it if it does.

We may also want to consider not allow different PRs to re-use tasks from other PRs (either by adjusting the index, or some sort of hard coded safeguard), although I'm not entirely sure it's necessary.

**Comment by bhearsum** [comment]
Some of RelEng talked about this over Zoom today. Here's an attempt as a summary:
- In the past, when using level 1 cached tasks was allowed in PRs, there were issues with different PRs stomping on each other.
  - This could be prevented by including the PR number in the cached task digest, or index route. In either case we'd still want to use level 3 caches first if they exist.
- Due to the way scopes work, any PR can publish to any level 1 index. This means that this is a potential security issue to allow, particularly in places where `push` or `release` tasks also run on level 1 workers.

All of this means that this is not a trivial fix, and any changes most likely need an RRA.

**Comment by bhearsum** [comment]
For context and a different perspective, one of the folks working on https://github.com/mozilla/firefox-translations-training offered this as what he expected as a user:
> I would expect the following behavior: if all of the data and code files that serve as a step input (including ci kinds and transforms) are unchanged, it should use the cached output. Sometimes you want to debug the last step of the pipeline on taskcluster and CI is a convenient way to do that. Then it shouldn’t rerun the whole thing every time. It adds to costs and wait time. I’ll file an issue in the repo

**Comment by ahal** [comment]
I think my preference here is to bake the PR number into the digest of the cached task. This way there could only be a collision if someone is maliciously trying to poison a PR cache from another PR. It seems pretty unlikely and wouldn't accomplish anything besides confuse the author of the target PR.

I also don't think it wouldn't introduce any new risks for level 1 repos, as PRs can already maliciously insert things into the index over there. So I don't think we would need an RRA for this.

As Julien pointed out in the meeting, the main complexity will stem from us needing to generate two digests:
* One without the PR number so we can check if it exists on level 3
* One with the PR number so we can check if it exists on level 1

**Comment by ahal** [comment]
I decided to hit send anyway, but as I was typing that out I realized that I'm not sure that baking the PR number into the digest really buys us anything. If the digests are legitimately the same, then it shouldn't matter if PRs are using cached tasks from other PRs, as the artifacts are theoretically identical.

And as discussed in the last comment, baking the PR number doesn't protect against malicious insertion either. So it's only really protecting us from invalidation bugs.. which is maybe useful? But imo better to detect those bugs and fix them anyway.
